### PR TITLE
[#IC-117] Implement selective session resolution

### PR DESCRIPTION
**Pull request by balanza**
This PR introduces the env variable `IDP: azure-ad | selfcare` based on which the application select the strategy to use for session handling.

As default value is `azure-ad`, these changes will have no effect of deployed on the existing configuration

### CODE REVIEW

File: src/controllers/configuration.ts [hunk 1]
```diff
@@ -12,18 +12,31 @@ import * as config from "../config";
...
+      const idp: never = config.IDP;
+      throw new Error(`Invalid IDP: ${idp}`);
+  }
+})();
 
 export async function getConfiguration(
   _: express.Request
-): Promise<IResponseSuccessJson<typeof msalConfig> | IResponseErrorInternal> {
-  return ResponseSuccessJson(msalConfig);
+): Promise<IResponseSuccessJson<typeof publicConfig> | IResponseErrorInternal> {
```
Comment (BurnedMarshal) [inline src/controllers/configuration.ts:40]: This functions returns always `ResponseSuccessJson`.
```suggestion
): Promise<IResponseSuccessJson<typeof publicConfig>> {
```

File: tsconfig.json [hunk 2]
```diff
@@ -11,7 +11,8 @@
     "noFallthroughCasesInSwitch": true,
     "resolveJsonModule": true,
     "sourceMap": true,
-    "outDir": "build"
+    "outDir": "build",
+    "skipLibCheck": true
```
Comment (BurnedMarshal) [inline tsconfig.json:15]: Why is this needed?

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: src/auth-strategies/selfcare_session_strategy.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+export type SelfCareUser = t.TypeOf<typeof SelfCareUser>;
+export const SelfCareUser = t.interface({
+  family_name: t.string,
+  given_name: t.string,
+  oid: NonEmptyString,
+  organization: t.interface({
+    id: OrganizationFiscalCode,
+    role: NonEmptyString
+  })
+});
```
[the remark this answers is not in the document]
Comment (balanza) [reply src/auth-strategies/selfcare_session_strategy.ts:25]: Not at the moment

File: tsconfig.json (hunk 2 above)
Comment (balanza) [reply tsconfig.json:15]: We have errors coming from type definitions for newly-introduced packages:
```sh
node_modules/@types/passport-strategy/node_modules/@types/passport/index.d.ts:23:13 - error TS2717: Subsequent property declarations must have the same type.  Property 'authInfo' must be of type 'any', but here has type 'AuthInfo | undefined'.

23             authInfo?: AuthInfo | undefined;
               ~~~~~~~~

  node_modules/@types/passport/index.d.ts:14:13
    14             authInfo?: any;
                   ~~~~~~~~
    'authInfo' was also declared here.

node_modules/@types/passport-strategy/node_modules/@types/passport/index.d.ts:24:13 - error TS2717: Subsequent property declarations must have the same type.  Property 'user' must be of type 'any', but here has type 'User | undefined'.

24             user?: User | undefined;
               ~~~~

  node_modules/@types/passport/index.d.ts:15:13
    15             user?: any;
                   ~~~~
    'user' was also declared here.
```

I think this is because we build in strict mode, but maybe such definitions aren't done fort

**Review by BurnedMarshal — APPROVED** [review]
LGTM, there is a small error in pipeline for an unused import.

**Review by BurnedMarshal — APPROVED** [review]
